reqresp/workflow.go

```diff
@@ -0,0 +1,195 @@
+// UppercaseWorkflow.
+type Request struct {
+	// ID of the request, also set on the response.
+	ID string `json:"id"`
+	// String to be uppercased.
+	Input string `json:"input"`
+
+	// If these two values are set, the Response is sent as a single param to the
+	// given activity on the given task queue. Failure to send to the activity is
+	// logged but does not fail the workflow.
+	ResponseActivity  string `json:"response_activity"`
+	ResponseTaskQueue string `json:"response_task_queue"`
+}
+
+// Response is a response to a Request. This can be returned in a map from a
+// query or as the parameter of a callback to a response activity.
+type Response struct {
+	ID     string `json:"id"`
+	Output string `json:"output"`
+	Error  string `json:"error"`
+}
+
+type uppercaser struct {
+	workflow.Context
+	requestCh                   workflow.ReceiveChannel
+	requestsBeforeContinueAsNew int
+	// This maintains responses for the lifetime of the workflow, which should not
+	// be too large before continue-as-new. This map is only here to serve queries
+	// and could be removed in a response-activity-only setup.
+	responses               map[string]*Response
+	responseActivityOptions workflow.ActivityOptions
+}
+
+func newUppercaser(ctx workflow.Context) (*uppercaser, error) {
+	u := &uppercaser{
+		// For the main context, we're only going to allow 1 retry and only a 5
+		// second schedule-to-close timeout. WARNING: The timeout and retry affect
+		// how long this workflow stays open and may prevent it from performing its
+		// continue-as-new until timeout occurs and/or retries are finished.
+		Context: workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+			ScheduleToCloseTimeout: 5 * time.Second,
+			RetryPolicy:            &temporal.RetryPolicy{MaximumAttempts: 2},
+		}),
+
+		// Use a signal to obtain requests. Since this cannot be closed, the
+		// workflow counts on the ability to have some "idle" period between
+		// handling requests/responses where it can continue-as-new.
+		requestCh: workflow.GetSignalChannel(ctx, "request"),
+
+		// We'll allow 500 requests before we continue-as-new the workflow. This is
+		// required because the history will grow very large otherwise for an
+		// interminable workflow fielding signal requests and executing activities.
+		requestsBeforeContinueAsNew: 500,
+		responses:                   make(map[string]*Response, 500),
+
+		// We're gonna just support 10 second timeout and 3 retries on response
+		// callback activities. WARNING: The timeout and retry affect how long this
+		// workflow stays open and may prevent it from performing its
+		// continue-as-new until timeout occurs and/or retries are finished.
+		responseActivityOptions: workflow.ActivityOptions{
+			// We use schedule-to-close because if the requester side is not present,
+			// this may hang otherwise with just start-to-close.
+			ScheduleToCloseTimeout: 10 * time.Second,
```

Comment:
Consider setting a short scheduleToStart to handle requester not present situation.

Reply:
I figured this setting was enough for that too since the value is so short and I expect the start-to-close to be very short. But I can set a schedule-to-start just for this demonstration just in case people repurposed the pattern with heavy response-handling activities.

Reply:
Added a `ScheduleToStart` that is only 5 seconds. I kept `ScheduleToClose` because I care about a completion time limit too.